### ISSUE TITLE
HIP: array-valued parallel_reduce fails in debug builds

### ISSUE THREAD

**Issue by brian-kelley** [kokkos/kokkos#3638]
On HIP, reductions with runtime-sized arrays cause an error on device. This is the case where the functor has no ``final(value_type)`` member function, and the result is a view passed as the 3rd argument of parallel_reduce.

Here's a full program that replicates this on Caraway (rocm 3.8.0, and only for ``CMAKE_BUILD_TYPE=Debug``):
```
#include "Kokkos_Core.hpp"

using ExecSpace = Kokkos::DefaultExecutionSpace;
using MemSpace = typename ExecSpace::memory_space;
using IntMatrix = Kokkos::View<int**, MemSpace>;

struct ColumnSumFunctor
{
  typedef int value_type[];

  IntMatrix m_x;  //Matrix - will sum the values in each column
  int value_count;

  ColumnSumFunctor(const IntMatrix& x)
  {
    m_x = x;
    value_count = m_x.extent(1);
  }

  KOKKOS_INLINE_FUNCTION
  void operator() (const int& i, value_type sum) const
  {
    for (int k = 0; k < value_count; ++k) {
      sum[k] += m_x(i,k);
    }   
  }

  KOKKOS_INLINE_FUNCTION void init (volatile value_type update) const
  {
    for (int k = 0; k < value_count; ++k) {
      update[k] = 0;
    }   
  }

  KOKKOS_INLINE_FUNCTION void
  join (volatile value_type update,
        const volatile value_type source) const
  {
    for (int k = 0; k < value_count; ++k) {
      update[k] += source[k];
    }   
  }
};

int main()
{
  Kokkos::initialize();
  {
    std::cout << "Running int[] valued reducer on device " << ExecSpace::name() << "/" << MemSpace::name() << '\n';
    IntMatrix x("Int matrix", 1000, 10);
    auto xhost = Kokkos::create_mirror_view(x);
    int m = 1000;
    int n = 10;
    for(int i = 0; i < m; i++)
    {
      for(int j = 0; j < n; j++)
      {
        xhost(i, j) = (i + j) % 5;
      }
    }
    Kokkos::deep_copy(x, xhost);
    Kokkos::View<int*, Kokkos::HostSpace> sums("Sums", n);
    Kokkos::View<int*, MemSpace> sumsDevice("Sums", n);
    Kokkos::parallel_reduce(Kokkos::RangePolicy<ExecSpace>(0, m), ColumnSumFunctor(x), sumsDevice);
    Kokkos::deep_copy(sums, sumsDevice);
    for(int i = 0; i < n; i++)
    {
      std::cout << sums(i) << ' ';
    }
    std::cout << '\n';
  }
  Kokkos::finalize();
}
```

On CUDA, this works:
```
[bmkelle@s1042556 blas]$ ./ArrayReduce 
Running int[] valued reducer on device Cuda/Cuda
2000 2000 2000 2000 2000 2000 2000 2000 2000 2000 
```
but on HIP:
```
Running int[] valued reducer on device HIP/HIP
:0:rocdevice.cpp            :2158: 601186817149 us: Device::callbackQueue aborting with status: 0x100f
Aborted
```

The behavior is the same if the result view passed to parallel_reduce is in HostSpace instead (still works on Cuda, still fails on HIP).

This affects the KokkosKernels rank-2 BLAS functions like dot, nrm1, iamax, etc.

**Comment by dalg24** [comment]
This is interesting.  This behavior only occurs in `Debug`.  I cannot reproduce with `Release` nor `RelWithDebInfo`.  I tried both ROCm 3.8 and 3.9

**Comment by Rombur** [comment]
I'll take a look but I have seen this error many times and it's usually a compiler bug. The fact that it only happens in `Debug` makes me think that it's once again a compiler bug.

**Comment by Rombur** [comment]
You are hitting a compiler bug that's been fix in the development version of the compiler. I don't know if rocm 3.10 works but master from two weeks ago does.

**Comment by Rombur** [comment]
I just tested with rocm 3.10 and there is still a problem. I guess you need to wait for the next release (or use master).

**Comment by dalg24** [comment]
This is still not fixed in ROCm 4.0

**Comment by Rombur** [comment]
The test passes with ROCm 4.1
